src/main/java/com/openlattice/datastore/data/controllers/DataController.java

```diff
@@ -309,24 +310,44 @@ public Integer replaceEntityProperties(
+        final var srcAssociationEntitySetIds = new HashMap<UUID, Set<UUID>>(); // edge-src
+        final var dstAssociationEntitySetIds = new HashMap<UUID, Set<UUID>>(); // edge-dst
+
+        final var entitySetIdChecks = new HashMap<AclKey, EnumSet<Permission>>();
+        final SetMultimap<UUID, UUID> keyPropertiesOfAssociations = HashMultimap.create();
         associations.forEach(
                 association -> {
-                    UUID associationEntitySetId = association.getEdge().getEntitySetId();
-
-                    //Ensure that we have read access to entity set metadata.
-                    ensureReadAccess( new AclKey( association.getSrc().getEntitySetId() ) );
-                    ensureReadAccess( new AclKey( association.getDst().getEntitySetId() ) );
-                    ensureReadAccess( new AclKey( associationEntitySetId ) );
-
-                    //Ensure that we can read key properties.
-                    Set<UUID> keyPropertyTypes = edmService
-                            .getEntityTypeByEntitySetId( associationEntitySetId ).getKey();
-                    keyPropertyTypes.forEach( propertyType ->
-                            accessCheck( new AclKey( associationEntitySetId, propertyType ), READ_PERMISSION ) );
+                    final var edgeEntitySetId = association.getEdge().getEntitySetId();
+                    final var srcEntitySetId = association.getSrc().getEntitySetId();
+                    final var dstEntitySetId = association.getDst().getEntitySetId();
+
+                    entitySetIdChecks.put( new AclKey( edgeEntitySetId ), READ_PERMISSION );
+                    entitySetIdChecks.put( new AclKey( srcEntitySetId ), READ_PERMISSION );
+                    entitySetIdChecks.put( new AclKey( dstEntitySetId ), READ_PERMISSION );
+
+                    final var keyPropertyTypes = edmService.getEntityTypeByEntitySetId( edgeEntitySetId ).getKey();
+                    keyPropertiesOfAssociations.putAll( edgeEntitySetId, keyPropertyTypes );
+
+                    if ( srcAssociationEntitySetIds
+                            .putIfAbsent( edgeEntitySetId, Sets.newHashSet( srcEntitySetId ) ) != null ) {
+                        srcAssociationEntitySetIds.get( edgeEntitySetId ).add( srcEntitySetId );
+                    }
+
+                    if ( dstAssociationEntitySetIds
+                            .putIfAbsent( edgeEntitySetId, Sets.newHashSet( dstEntitySetId ) ) != null ) {
+                        dstAssociationEntitySetIds.get( edgeEntitySetId ).add( dstEntitySetId );
+                    }
                 }
         );
 
-        WriteEvent writeEvent = dgm.createAssociations( associations );
+        //Ensure that we have read access to entity set metadata.
+        accessCheck( entitySetIdChecks );
+
+        //Ensure that we can read key properties on association.
+        accessCheck( aclKeysForAccessCheck( keyPropertiesOfAssociations, READ_PERMISSION ) );
```

Comment:
Do you know why this is only checking key property types? (I see this is how we were doing it before but not sure what the logic is there) (also unclear to me why it was/is checking read and not write)